File: 20170508-typescript-file-logger/test/file-logger.spec.ts
import FileLogger from '../app/file-logger';
import FileSystem from '../app/file-system';

import { expect } from 'chai';
import * as sinon from 'sinon';

describe('FileLogger', () => {

    let sandbox = sinon.sandbox.create();

    afterEach(function () {
        sandbox.restore();
    });

    describe('log()', () => {

        it('should append message to log.txt file', () => {
            const message = 'New message';
            const appendSpy = sandbox.stub(FileSystem.prototype, 'append');
            const fileLogger = new FileLogger(new FileSystem());
            sandbox.stub(FileSystem.prototype, 'exists').returns(true);

            fileLogger.log(message);

            expect(appendSpy.calledWith('log.txt', message)).to.be.true;
        });

        it('should write message to a new log.txt file if it does not exist yet', () => {
            const message = 'New message';
            const writeSpy = sandbox.stub(FileSystem.prototype, 'write');
            const fileLogger = new FileLogger(new FileSystem());
            sandbox.stub(FileSystem.prototype, 'exists').returns(false);

            fileLogger.log(message);

            expect(writeSpy.calledWith('log.txt', message)).to.be.true;
        });

    });

});

File: 20171011-kotlin-file-logger/src/test/kotlin/com/katas/FileLoggerSpec.kt
package com.katas

import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on
import org.mockito.Mockito.*
import java.time.LocalDateTime

object FileLoggerSpec : Spek({

    describe("File Logger") {

        val message = "A message"
        val aSaturday = LocalDateTime.of(2017, 10, 14, 0, 0)
        val aSunday = LocalDateTime.of(2017, 10, 15, 0, 0)
        val aMonday = LocalDateTime.of(2017, 10, 16, 0, 0)

        val fileWriterMock = mock(FileWriter::class.java)
        val dateProviderMock = mock(DateProvider::class.java)
        val logger = FileLogger(fileWriterMock, dateProviderMock)

        it("should append message to end of file when file already exists for the day") {
            `when`(dateProviderMock.currentDate()).thenReturn(aMonday)
            `when`(fileWriterMock.exist("log20171016.txt")).thenReturn(true)

            logger.log(message)

            verify(fileWriterMock).appendText("log20171016.txt", message)
        }

        it("should write message to new file when file does not exist yet for the day") {
            `when`(dateProviderMock.currentDate()).thenReturn(aMonday)
            `when`(fileWriterMock.exist("log20171016.txt")).thenReturn(false)

            logger.log(message)

            verify(fileWriterMock).writeText("log20171016.txt", message)
        }

        on("Saturday or Sunday") {

            it("should append message to end of file when weekend.txt file already exists") {
                `when`(dateProviderMock.currentDate()).thenReturn(aSaturday)
                `when`(fileWriterMock.exist("weekend.txt")).thenReturn(true)

                logger.log(message)

                verify(fileWriterMock).appendText("weekend.txt", message)
            }

            it("should write message to new file when weekend.txt file does not exist yet") {
                `when`(dateProviderMock.currentDate()).thenReturn(aSunday)
                `when`(fileWriterMock.exist("weekend.txt")).thenReturn(false)

                logger.log(message)

                verify(fileWriterMock).writeText("weekend.txt", message)
            }

        }

    }

})